core/services/directrequestocr/orm.go

```diff
@@ -19,121 +19,90 @@ type ORM interface {
+		VALUES ($1,$2,$3,$4,$5);
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID, o.contractAdderss, receivedAt, requestTxHash, IN_PROGRESS)
+}
 
-	o.counter++
-	newEntry := Request{
-		ID:            o.counter,
-		RequestID:     requestID,
-		ReceivedAt:    receivedAt,
-		RequestTxHash: requestTxHash,
-		State:         IN_PROGRESS,
-	}
-	o.db[requestID] = newEntry
-	return nil
+func (o *orm) SetResult(requestID RequestID, runID int64, computationResult []byte, readyAt time.Time, qopts ...pg.QOpt) error {
+	stmt := `
+		UPDATE ocr2dr_requests
+		SET (run_id=$2, result=$3, result_ready_at=$4, state=$5)
+		WHERE request_id=$1;
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID, runID, computationResult, readyAt, RESULT_READY)
 }
 
-func (o *inmemoryorm) SetResult(requestID RequestID, runID int64, computationResult []byte, readyAt time.Time, qopts ...pg.QOpt) error {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		val.RunID = runID
-		val.ErrorType = NONE
-		val.Result = computationResult
-		val.Error = []byte{}
-		val.State = RESULT_READY
-		val.ResultReadyAt = readyAt
-		o.db[requestID] = val
-		return nil
-	}
-	return fmt.Errorf("can't find entry with requestID: %v", requestID)
+func (o *orm) SetError(requestID RequestID, runID int64, errorType ErrType, computationError []byte, readyAt time.Time, qopts ...pg.QOpt) error {
+	stmt := `
+		UPDATE ocr2dr_requests
+		SET (run_id=$2, error=$3, error_type=$4, result_ready_at=$5, state=$6)
+		WHERE request_id=$1;
+	`
+	return o.q.WithOpts(qopts...).ExecQ(stmt, requestID, runID, computationError, errorType, readyAt, RESULT_READY)
 }
 
-func (o *inmemoryorm) SetError(requestID RequestID, runID int64, errorType ErrType, computationError []byte, readyAt time.Time, qopts ...pg.QOpt) error {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	if val, ok := o.db[requestID]; ok {
-		val.RunID = runID
-		val.ErrorType = errorType
-		val.Error = computationError
-		val.State = RESULT_READY
-		val.Result = []byte{}
-		val.ResultReadyAt = readyAt
-		o.db[requestID] = val
-		return nil
+func (o *orm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
+	var oldState RequestState
+
+	stmt := `SELECT state FROM ocr2dr_requests WHERE request_id=$1;`
+	if err := o.q.WithOpts(qopts...).Get(&oldState, stmt); err != nil {
+		return 0, err
 	}
-	return fmt.Errorf("can't find entry with requestID: %v", requestID)
-}
 
-func (o *inmemoryorm) SetState(requestID RequestID, state RequestState, qopts ...pg.QOpt) (RequestState, error) {
-	o.mutex.Lock()
-	defer o.mutex.Unlock()
-	prevState := IN_PROGRESS
-	if val, ok := o.db[requestID]; ok {
-		prevState = val.State
-		val.State = state
-		o.db[requestID] = val
-		return prevState, nil
+	if oldState == state {
+		return state, nil
 	}
-	return prevState, fmt.Errorf("can't find entry with requestID: %v", requestID)
+
+	stmt = `UPDATE ocr2dr_requests SET (state=$2) WHERE request_id=$1;`
```

Comment:
Shouldn't SELECT and UPDATE happen in a single transaction?

Reply:
This statement should not return anything, hence no SELECT. But I will check with tests if I missed something.